In [1]:
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt

In [4]:
frame = pd.read_csv("./data/mammals.csv")
frame.head()

name     body  brain
0       Arctic fox    3.385   44.5
1       Owl monkey    0.480   15.5
2  Mountain beaver    1.350    8.1
3              Cow  465.000  423.0
4        Grey wolf   36.330  119.5

# Modyfikacja danych
## .rename()
### Zmiana nazw kolumn
Żeby zmienic nazwę kolumn musimy do .columns przypisać listę z nowymi nazwami wszystkich kolumn 

In [7]:
frame.columns = ['name', 'bodyKg', 'brainKg']

In [8]:
frame.head()

name   bodyKg  brainKg
0       Arctic fox    3.385     44.5
1       Owl monkey    0.480     15.5
2  Mountain beaver    1.350      8.1
3              Cow  465.000    423.0
4        Grey wolf   36.330    119.5

Żeby zmienić tylko jedną nazwę kolumny musimy użyć słownika {'stara nazwa':'nowa nazwa',{'stara nazwa':'nowa nazwa'},
<br> następnie przez funkcję .rename(columns = słownik) przypisać nowe nazwy

In [9]:
newColumnNames = {'bodyKg':'body_kg', 'brainKg':'brain_kg'}

In [10]:
frame.rename(columns = newColumnNames, inplace = True)

In [ ]:
frame.rename(columns = {'bodyKg':'body_kg', 'brainKg':'brain_kg'}, inplace = True) # inny zapis bez zmiennej

### Zmiana nazw indexów
.rename(index = słownik, inplace = True)

In [12]:
frame = pd.read_csv("./data/mammals.csv", index_col='name')

In [13]:
frame.rename(index = {'Cow':'Land cow'}, inplace = True)

In [14]:
frame.head()

body  brain
name                           
Arctic fox         3.385   44.5
Owl monkey         0.480   15.5
Mountain beaver    1.350    8.1
Land cow         465.000  423.0
Grey wolf         36.330  119.5

### .copy()
### Tworzenie kopii dataFrame
Trzeba przypisać do nowej zmiennej dotychczasowy dataFrame z funkcją .copy()

In [15]:
frameCopy = frame.copy()

zmiana w dancyh

In [16]:
frameCopy.loc['Grey wolf','body'] = 40      # zmieniamy w wierszu Grey wolf i kolumnie body wartość na 40

In [17]:
frameCopy.head()

body  brain
name                           
Arctic fox         3.385   44.5
Owl monkey         0.480   15.5
Mountain beaver    1.350    8.1
Land cow         465.000  423.0
Grey wolf         40.000  119.5

In [18]:
frame.head()

body  brain
name                           
Arctic fox         3.385   44.5
Owl monkey         0.480   15.5
Mountain beaver    1.350    8.1
Land cow         465.000  423.0
Grey wolf         36.330  119.5

## Ćwiczenie
Edycja kolumn i indexu

In [19]:
import pandas as pd

In [20]:
professions = pd.read_csv('./data/Prestige.csv', index_col='name')

In [21]:
professionsOriginal = professions.copy()

In [22]:
professionsOriginal.head()

education  income  women  prestige  census  type
name                                                                 
gov.administrators       13.11   12351  11.16      68.8    1113  prof
general.managers         12.26   25879   4.02      69.1    1130  prof
accountants              12.77    9271  15.70      63.4    1171  prof
purchasing.officers      11.42    8865   9.11      56.8    1175  prof
chemists                 14.62    8403  11.68      73.5    2111  prof

In [25]:
professions.rename(columns = {'income':'IncomeDollarsPerYear'}, inplace=True)

In [26]:
professions.head()

education  IncomeDollarsPerYear  women  prestige  census  \
name                                                                            
gov.administrators       13.11                 12351  11.16      68.8    1113   
general.managers         12.26                 25879   4.02      69.1    1130   
accountants              12.77                  9271  15.70      63.4    1171   
purchasing.officers      11.42                  8865   9.11      56.8    1175   
chemists                 14.62                  8403  11.68      73.5    2111   

                     type  
name                       
gov.administrators   prof  
general.managers     prof  
accountants          prof  
purchasing.officers  prof  
chemists             prof

In [27]:
professions.rename(columns = {'women':'womenPercent', 'prestige':'prestigeIndicator'}, inplace=True)

In [28]:
professions.head()

education  IncomeDollarsPerYear  womenPercent  \
name                                                                 
gov.administrators       13.11                 12351         11.16   
general.managers         12.26                 25879          4.02   
accountants              12.77                  9271         15.70   
purchasing.officers      11.42                  8865          9.11   
chemists                 14.62                  8403         11.68   

                     prestigeIndicator  census  type  
name                                                  
gov.administrators                68.8    1113  prof  
general.managers                  69.1    1130  prof  
accountants                       63.4    1171  prof  
purchasing.officers               56.8    1175  prof  
chemists                          73.5    2111  prof

In [29]:
professions.rename(index = {'accountants':'bookkeepers', 'general.managers':'managers'}, inplace=True)

In [30]:
professions.head()

education  IncomeDollarsPerYear  womenPercent  \
name                                                                 
gov.administrators       13.11                 12351         11.16   
managers                 12.26                 25879          4.02   
bookkeepers              12.77                  9271         15.70   
purchasing.officers      11.42                  8865          9.11   
chemists                 14.62                  8403         11.68   

                     prestigeIndicator  census  type  
name                                                  
gov.administrators                68.8    1113  prof  
managers                          69.1    1130  prof  
bookkeepers                       63.4    1171  prof  
purchasing.officers               56.8    1175  prof  
chemists                          73.5    2111  prof

In [31]:
professionsOriginal.head()

education  income  women  prestige  census  type
name                                                                 
gov.administrators       13.11   12351  11.16      68.8    1113  prof
general.managers         12.26   25879   4.02      69.1    1130  prof
accountants              12.77    9271  15.70      63.4    1171  prof
purchasing.officers      11.42    8865   9.11      56.8    1175  prof
chemists                 14.62    8403  11.68      73.5    2111  prof

## Modyfikacja wartości w dataFrame
.loc[wiesz, kolumna] , iloc[wiesz numerycznie, kolumna numerycznie]

In [32]:
frame = pd.read_csv('./data/insurance.csv')

In [33]:
frame.loc[2, 'Claims'] = 19

In [34]:
frame.loc[2, 'Claims'] += 1      # dodaj 1 do aktualnej wartości

Modyfikacja większej ilości wierszy 

In [35]:
isYounger25 = frame["Age"] == "<25"        # tam gdzie jest spełniony warunek otrzymamy wartość True

In [36]:
frame[isYounger25].head()              # wyświetli tylko te wartości, gdzie isYounger25 było True

ID  District   Group  Age  Holders  Claims
0    1         1     <1l  <25      197      38
4    5         1  1-1.5l  <25      284      63
8    9         1  1.5-2l  <25      133      19
12  13         1     >2l  <25       24       4
16  17         2     <1l  <25       85      22

In [37]:
frame.loc[isYounger25, "Holders"]      # wyświetli indexy i kolumnę Holders dla wartości, gdzie isYounger25 było True

0     197
4     284
8     133
12     24
16     85
20    149
24     66
28      9
32     35
36     53
40     24
44      7
48     20
52     31
56     18
60      3
Name: Holders, dtype: int64

In [39]:
frame.loc[isYounger25, "Holders"] + 100     # zmieniamy wszystkie wartości o +100

0     297
4     384
8     233
12    124
16    185
20    249
24    166
28    109
32    135
36    153
40    124
44    107
48    120
52    131
56    118
60    103
Name: Holders, dtype: int64

Przypisujemy zmianę do wybranych przez nas wartości na stałe

In [40]:
frame.loc[isYounger25, "Holders"] = frame.loc[isYounger25, "Holders"] + 100 

In [41]:
frame.loc[isYounger25, "Holders"] = frame["Holders"] + 100        # ten sam zapis co wiersz wyżej. Zmieni tylko tam gdzie było True

In [ ]:
frame.loc[isYounger25, "Holders"] -= 50    # zmniejszy na stałe wartości o 50

## Ćwiczenie 
Modyfikacja danych

In [42]:
import pandas as pd

In [52]:
professions = pd.read_csv('./data/Prestige.csv', index_col='name')

In [53]:
professions.loc['chemists', 'type'] = 'scientist'

In [54]:
professions.head()

education  income  women  prestige  census       type
name                                                                      
gov.administrators       13.11   12351  11.16      68.8    1113       prof
general.managers         12.26   25879   4.02      69.1    1130       prof
accountants              12.77    9271  15.70      63.4    1171       prof
purchasing.officers      11.42    8865   9.11      56.8    1175       prof
chemists                 14.62    8403  11.68      73.5    2111  scientist

In [55]:
isScientist = (professions['type'] == 'scientist')               # nawiasy bez nich tez działa - professions['type'] == 'scientist'

In [56]:
professions.loc[isScientist, 'prestige'] = 90

In [57]:
professions.head()

education  income  women  prestige  census       type
name                                                                      
gov.administrators       13.11   12351  11.16      68.8    1113       prof
general.managers         12.26   25879   4.02      69.1    1130       prof
accountants              12.77    9271  15.70      63.4    1171       prof
purchasing.officers      11.42    8865   9.11      56.8    1175       prof
chemists                 14.62    8403  11.68      90.0    2111  scientist

In [58]:
professions.loc[isScientist, 'income'] = professions['income'] * 1.5

In [59]:
professions.loc[isScientist, 'income'] *= 1.5        # szybszy zapis

In [60]:
professions.head()

education    income  women  prestige  census       type
name                                                                        
gov.administrators       13.11  12351.00  11.16      68.8    1113       prof
general.managers         12.26  25879.00   4.02      69.1    1130       prof
accountants              12.77   9271.00  15.70      63.4    1171       prof
purchasing.officers      11.42   8865.00   9.11      56.8    1175       prof
chemists                 14.62  18906.75  11.68      90.0    2111  scientist

## Dodawanie i usuwanie wierszy i kolumn

In [63]:
frame = pd.read_csv('./data/sleep_time.csv', index_col="ID")

In [64]:
frame.head()

name       genus   vore         order  conservation  \
ID                                                                              
1                      Cheetah    Acinonyx  carni     Carnivora            lc   
2                   Owl monkey       Aotus   omni      Primates           NaN   
3              Mountain beaver  Aplodontia  herbi      Rodentia            nt   
4   Greater short-tailed shrew     Blarina   omni  Soricomorpha            lc   
5                          Cow         Bos  herbi  Artiodactyla  domesticated   

    sleep_total  sleep_rem  sleep_cycle  awake  brainwt   bodywt  
ID                                                                
1          12.1        NaN          NaN   11.9      NaN   50.000  
2          17.0        1.8          NaN    7.0  0.01550    0.480  
3          14.4        2.4          NaN    9.6      NaN    1.350  
4          14.9        2.3     0.133333    9.1  0.00029    0.019  
5           4.0        0.7     0.666667   20.0  0.42300  600.000

### Usuwanie
#### del i .drop()

In [65]:
del frame["bodywt"]             # usuwa kolumnę bodywt

In [ ]:
frame.drop(axis=1, columns=['awake', 'brainwt'], inplcae=True)           # usuwa kilka kolumn (axis=1 == axis='columns')

In [5]:
frame.drop(axis='rows', labels=[1, 2], inplcae=True)                  # usuwa wiersze (axis='rows' == axis=0)

<class 'NameError'>: name 'frame' is not defined

skrócone wersje

In [67]:
frame.drop(columns=['awake', 'brainwt'], inplace=True)      #usuwa kolumny

In [ ]:
frame.drop(labels=[1, 2], inplace=True)         # usuwa wiersze, lub jeszcze prościej...

In [68]:
frame.drop([1, 2], inplace=True) 

In [69]:
frame.head()

name        genus   vore         order  \
ID                                                                 
3              Mountain beaver   Aplodontia  herbi      Rodentia   
4   Greater short-tailed shrew      Blarina   omni  Soricomorpha   
5                          Cow          Bos  herbi  Artiodactyla   
6             Three-toed sloth     Bradypus  herbi        Pilosa   
7            Northern fur seal  Callorhinus  carni     Carnivora   

    conservation  sleep_total  sleep_rem  sleep_cycle  
ID                                                     
3             nt         14.4        2.4          NaN  
4             lc         14.9        2.3     0.133333  
5   domesticated          4.0        0.7     0.666667  
6            NaN         14.4        2.2     0.766667  
7             vu          8.7        1.4     0.383333

### Dodawanie 
### .append({'kolumna':'wartość', 'kolumna':'wartość', 'kolumna':'wartość', 'kolumna':'wartość'})
<br> wartości poprzez append są dodawane na końcu dataFrame

możemy dodać większą ilość elementów jednocześnie 

In [71]:
subset = frame[0:5]         # bierzemy elementy od 0 do 4 i przypisujemy do zmiennej subset

In [72]:
frame.append(subset)

C:\Users\hp\AppData\Local\Temp\ipykernel_4448\3439080439.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  frame.append(subset)


name        genus   vore         order  \
ID                                                                 
3              Mountain beaver   Aplodontia  herbi      Rodentia   
4   Greater short-tailed shrew      Blarina   omni  Soricomorpha   
5                          Cow          Bos  herbi  Artiodactyla   
6             Three-toed sloth     Bradypus  herbi        Pilosa   
7            Northern fur seal  Callorhinus  carni     Carnivora   
..                         ...          ...    ...           ...   
3              Mountain beaver   Aplodontia  herbi      Rodentia   
4   Greater short-tailed shrew      Blarina   omni  Soricomorpha   
5                          Cow          Bos  herbi  Artiodactyla   
6             Three-toed sloth     Bradypus  herbi        Pilosa   
7            Northern fur seal  Callorhinus  carni     Carnivora   

    conservation  sleep_total  sleep_rem  sleep_cycle  
ID                                                     
3             nt         14.4        2.4          NaN  
4             lc         14.9        2.3     0.133333  
5   domesticated          4.0        0.7     0.666667  
6            NaN         14.4        2.2     0.766667  
7             vu          8.7        1.4     0.383333  
..           ...          ...        ...          ...  
3             nt         14.4        2.4          NaN  
4             lc         14.9        2.3     0.133333  
5   domesticated          4.0        0.7     0.666667  
6            NaN         14.4        2.2     0.766667  
7             vu          8.7        1.4     0.383333  

[86 rows x 8 columns]

Żeby zachować dodane wartości na stałe musimy przypisać działanie do zmiennej

In [ ]:
frame = frame.append(subset)

## Ćwiczenie 
Dodawanie i usuwanie wierszy

In [73]:
import pandas as pd

In [74]:
professions = pd.read_csv('./data/Prestige.csv')

In [80]:
dict = { 'name' : 'data scientist',
'education' : np.NaN,
'income' : np.NaN,
'women' : np.NaN,
'prestige' : np.NaN,
'census' : np.NaN,
'type' : np.NaN}

In [91]:
professions.append(dict, ignore_index=True)

C:\Users\hp\AppData\Local\Temp\ipykernel_4448\4114828814.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  professions.append(dict, ignore_index=True)


name  education  income  women  prestige  census type
97    bus.drivers       7.58    5562   9.47      35.9    9171   bc
98   taxi.drivers       7.93    4224   3.59      25.1    9173   bc
99   longshoremen       8.37    4753   0.00      26.1    9313   bc
100   typesetters      10.00    6462  13.58      42.2    9511   bc
101   bookbinders       8.55    3617  70.87      35.2    9517   bc

In [92]:
professions.tail()

name  education  income  women  prestige  census type
97    bus.drivers       7.58    5562   9.47      35.9    9171   bc
98   taxi.drivers       7.93    4224   3.59      25.1    9173   bc
99   longshoremen       8.37    4753   0.00      26.1    9313   bc
100   typesetters      10.00    6462  13.58      42.2    9511   bc
101   bookbinders       8.55    3617  70.87      35.2    9517   bc

In [93]:
taxi = professions[professions['name'] == 'taxi.drivers']           # szuka w kolumnie name wartości taxi.drivers

In [94]:
taxi

name  education  income  women  prestige  census type
98  taxi.drivers       7.93    4224   3.59      25.1    9173   bc

In [95]:
professions.drop(98, inplace=True)

In [96]:
professions.tail()

name  education  income  women  prestige  census type
96   train.engineers       8.49    8845   0.00      48.9    9131   bc
97       bus.drivers       7.58    5562   9.47      35.9    9171   bc
99      longshoremen       8.37    4753   0.00      26.1    9313   bc
100      typesetters      10.00    6462  13.58      42.2    9511   bc
101      bookbinders       8.55    3617  70.87      35.2    9517   bc

In [97]:
professions = professions.append(taxi)

C:\Users\hp\AppData\Local\Temp\ipykernel_4448\113176898.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  professions = professions.append(taxi)


In [98]:
professions.tail()

name  education  income  women  prestige  census type
97    bus.drivers       7.58    5562   9.47      35.9    9171   bc
99   longshoremen       8.37    4753   0.00      26.1    9313   bc
100   typesetters      10.00    6462  13.58      42.2    9511   bc
101   bookbinders       8.55    3617  70.87      35.2    9517   bc
98   taxi.drivers       7.93    4224   3.59      25.1    9173   bc

In [99]:
professions.drop(columns='census', inplace=True)             

In [100]:
del professions["type"]

In [101]:
professions.head()

name  education  income  women  prestige
0   gov.administrators      13.11   12351  11.16      68.8
1     general.managers      12.26   25879   4.02      69.1
2          accountants      12.77    9271  15.70      63.4
3  purchasing.officers      11.42    8865   9.11      56.8
4             chemists      14.62    8403  11.68      73.5

## Przebudowa indexu
### .set_index(), reset_index()

In [102]:
frame = pd.read_csv("./data/mammals.csv")

In [106]:
frame.set_index('name', inplace=True)               # Ustawiamy jako index kolumnę name

In [107]:
frame.head()

body  brain
name                           
Arctic fox         3.385   44.5
Owl monkey         0.480   15.5
Mountain beaver    1.350    8.1
Cow              465.000  423.0
Grey wolf         36.330  119.5

przywracanie indexu do pierwotnej postaci

In [108]:
frame.reset_index(inplace=True)

ustawianie set_index() za każdym razem usuwa dotychczasowy index, przez co łatwo stracić dane. Zawsze przed zmianą indexu używamy reset_index(), żeby tracić domyślny index, zamiast takiego z danymi.

## Ćwiczenie 
Przebudowa indexu

In [109]:
import pandas as pd

In [111]:
fortune = pd.read_csv('./data/Fortune_500_2017.csv')

In [112]:
fortune.head()

Rank               Title                           Website  Employees  \
0     1             Walmart            http://www.walmart.com    2300000   
1     2  Berkshire Hathaway  http://www.berkshirehathaway.com     367700   
2     3               Apple              http://www.apple.com     116000   
3     4         Exxon Mobil         http://www.exxonmobil.com      72700   
4     5            McKesson           http://www.mckesson.com      68000   

        Sector                                  Industry         Hqlocation  \
0    Retailing                     General Merchandisers    Bentonville, AR   
1   Financials  Insurance: Property and Casualty (Stock)          Omaha, NE   
2   Technology               Computers, Office Equipment      Cupertino, CA   
3       Energy                        Petroleum Refining         Irving, TX   
4  Wholesalers                  Wholesalers: Health Care  San Francisco, CA   

                   Hqaddr         Hqcity Hqstate  ...  \
0     702 S.W. Eighth St.    Bentonville      AR  ...   
1         3555 Farnam St.          Omaha      NE  ...   
2         1 Infinite Loop      Cupertino      CA  ...   
3  5959 Las Colinas Blvd.         Irving      TX  ...   
4              1 Post St.  San Francisco      CA  ...   

                                       Ceo-title  \
0  President, Chief Executive Officer & Director   
1             Chairman & Chief Executive Officer   
2             Chief Executive Officer & Director   
3             Chairman & Chief Executive Officer   
4  Chairman, President & Chief Executive Officer   

                                      Address Ticker                 Fullname  \
0  702 S.W. Eighth St., Bentonville, AR 72716    WMT    Wal-Mart Stores, Inc.   
1            3555 Farnam St., Omaha, NE 68131   BRKA  Berkshire Hathaway Inc.   
2        1 Infinite Loop, Cupertino, CA 95014   AAPL              Apple, Inc.   
3    5959 Las Colinas Blvd., Irving, TX 75039    XOM  Exxon Mobil Corporation   
4         1 Post St., San Francisco, CA 94104    MCK     McKesson Corporation   

  Revenues Revchange  Profits  Prftchange  Assets  Totshequity  
0   485873       0.8  13643.0        -7.2  198825      77798.0  
1   223604       6.1  24074.0           0  620854     283001.0  
2   215639      -7.7  45687.0       -14.4  321686     128249.0  
3   205004     -16.7   7840.0       -51.5  330314     167325.0  
4   192487       6.2   2258.0          53   56563       8924.0  

[5 rows x 23 columns]

In [113]:
fortune.set_index('Rank', inplace=True)

In [114]:
fortune.loc[100:105]               # bez loc ? 

Title                    Website  Employees  \
Rank                                                                      
100         Capital One Financial  http://www.capitalone.com      47300   
101      Twenty-First Century Fox        http://www.21cf.com      21500   
102                          USAA        http://www.usaa.com      29943   
103           World Fuel Services     http://www.wfscorp.com       5000   
104   Philip Morris International         http://www.pmi.com      79500   
105                         Deere   http://www.johndeere.com      56767   

                         Sector                                  Industry  \
Rank                                                                        
100                  Financials                          Commercial Banks   
101                       Media                             Entertainment   
102                  Financials  Insurance: Property and Casualty (Stock)   
103                      Energy                                    Energy   
104   Food, Beverages & Tobacco                                   Tobacco   
105                 Industrials           Construction and Farm Machinery   

           Hqlocation                     Hqaddr       Hqcity Hqstate  Hqzip  \
Rank                                                                           
100        McLean, VA       1680 Capital One Dr.       McLean      VA  22102   
101      New York, NY  1211 Ave. of the Americas     New York      NY  10036   
102   San Antonio, TX    9800 Fredericksburg Rd.  San Antonio      TX  78288   
103         Miami, FL         9800 N.W. 41st St.        Miami      FL  33178   
104      New York, NY              120 Park Ave.     New York      NY  10017   
105        Moline, IL         1 John Deere Place       Moline      IL  61265   

      ...                                      Ceo-title  \
Rank  ...                                                  
100   ...  Chairman, President & Chief Executive Officer   
101   ...             Chief Executive Officer & Director   
102   ...  President, Chief Executive Officer & Director   
103   ...  Chairman, President & Chief Executive Officer   
104   ...             Chief Executive Officer & Director   
105   ...             Chairman & Chief Executive Officer   

                                             Address Ticker  \
Rank                                                          
100           1680 Capital One Dr., McLean, VA 22102    COF   
101    1211 Ave. of the Americas, New York, NY 10036   FOXA   
102   9800 Fredericksburg Rd., San Antonio, TX 78288   FOXA   
103              9800 N.W. 41st St., Miami, FL 33178    INT   
104                120 Park Ave., New York, NY 10017     PM   
105             1 John Deere Place, Moline, IL 61265     DE   

                                    Fullname Revenues Revchange  Profits  \
Rank                                                                       
100        Capital One Financial Corporation    27519       9.6   3751.0   
101           Twenty-First Century Fox, Inc.    27326      -5.7   2755.0   
102   United Services Automobile Association    27131      11.4   1779.1   
103          World Fuel Services Corporation    27016     -11.1    126.5   
104         Philip Morris International Inc.    26685      -0.4   6967.0   
105                          Deere & Company    26644      -7.7   1523.9   

      Prftchange  Assets Totshequity  
Rank                                  
100         -7.4  357033     47514.0  
101        -66.8   48365     13661.0  
102        -21.7  147290     28840.0  
103        -27.5    5413      1925.0  
104          1.4   36851    -12688.0  
105        -21.4   57981      6520.0  

[6 rows x 22 columns]

In [115]:
fortune.loc[100:105][["Title","Employees"]]

Title  Employees
Rank                                        
100         Capital One Financial      47300
101      Twenty-First Century Fox      21500
102                          USAA      29943
103           World Fuel Services       5000
104   Philip Morris International      79500
105                         Deere      56767

In [116]:
fortune.reset_index(inplace=True)

In [117]:
fortune.set_index('Title', inplace=True)

In [118]:
fortune.head()

Rank                           Website  Employees  \
Title                                                                   
Walmart                1            http://www.walmart.com    2300000   
Berkshire Hathaway     2  http://www.berkshirehathaway.com     367700   
Apple                  3              http://www.apple.com     116000   
Exxon Mobil            4         http://www.exxonmobil.com      72700   
McKesson               5           http://www.mckesson.com      68000   

                         Sector                                  Industry  \
Title                                                                       
Walmart               Retailing                     General Merchandisers   
Berkshire Hathaway   Financials  Insurance: Property and Casualty (Stock)   
Apple                Technology               Computers, Office Equipment   
Exxon Mobil              Energy                        Petroleum Refining   
McKesson            Wholesalers                  Wholesalers: Health Care   

                           Hqlocation                  Hqaddr         Hqcity  \
Title                                                                          
Walmart               Bentonville, AR     702 S.W. Eighth St.    Bentonville   
Berkshire Hathaway          Omaha, NE         3555 Farnam St.          Omaha   
Apple                   Cupertino, CA         1 Infinite Loop      Cupertino   
Exxon Mobil                Irving, TX  5959 Las Colinas Blvd.         Irving   
McKesson            San Francisco, CA              1 Post St.  San Francisco   

                   Hqstate  Hqzip  ...  \
Title                              ...   
Walmart                 AR  72716  ...   
Berkshire Hathaway      NE  68131  ...   
Apple                   CA  95014  ...   
Exxon Mobil             TX  75039  ...   
McKesson                CA  94104  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   
Apple                          Chief Executive Officer & Director   
Exxon Mobil                    Chairman & Chief Executive Officer   
McKesson            Chairman, President & Chief Executive Officer   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                     1 Infinite Loop, Cupertino, CA 95014   AAPL   
Exxon Mobil           5959 Las Colinas Blvd., Irving, TX 75039    XOM   
McKesson                   1 Post St., San Francisco, CA 94104    MCK   

                                   Fullname Revenues Revchange  Profits  \
Title                                                                     
Walmart               Wal-Mart Stores, Inc.   485873       0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604       6.1  24074.0   
Apple                           Apple, Inc.   215639      -7.7  45687.0   
Exxon Mobil         Exxon Mobil Corporation   205004     -16.7   7840.0   
McKesson               McKesson Corporation   192487       6.2   2258.0   

                    Prftchange  Assets Totshequity  
Title                                               
Walmart                   -7.2  198825     77798.0  
Berkshire Hathaway           0  620854    283001.0  
Apple                    -14.4  321686    128249.0  
Exxon Mobil              -51.5  330314    167325.0  
McKesson                    53   56563      8924.0  

[5 rows x 22 columns]

In [119]:
fortune.sort_index(inplace=True)

In [120]:
fortune.head()

Rank               Website  Employees  \
Title                                                           
3M                        94     http://www.3m.com      91584   
A-Mark Precious Metals   395  http://www.amark.com         83   
ABM Industries           500    http://www.abm.com     110000   
ADP                      240    http://www.adp.com      57000   
AECOM                    161  http://www.aecom.com      87000   

                                            Sector  \
Title                                                
3M                                     Industrials   
A-Mark Precious Metals                   Materials   
ABM Industries                   Business Services   
ADP                              Business Services   
AECOM                   Engineering & Construction   

                                                Industry        Hqlocation  \
Title                                                                        
3M                                         Miscellaneous      St. Paul, MN   
A-Mark Precious Metals                     Miscellaneous  Santa Monica, CA   
ABM Industries          Diversified Outsourcing Services      New York, NY   
ADP                     Diversified Outsourcing Services      Roseland, NJ   
AECOM                          Engineering, Construction   Los Angeles, CA   

                                        Hqaddr        Hqcity Hqstate  Hqzip  \
Title                                                                         
3M                                   3M Center      St. Paul      MN  55144   
A-Mark Precious Metals  429 Santa Monica Blvd.  Santa Monica      CA  90401   
ABM Industries                 1 Liberty Plaza      New York      NY  10006   
ADP                                1 ADP Blvd.      Roseland      NJ   7068   
AECOM                   1999 Ave. of the Stars   Los Angeles      CA  90067   

                        ...                                      Ceo-title  \
Title                   ...                                                  
3M                      ...  Chairman, President & Chief Executive Officer   
A-Mark Precious Metals  ...             Chief Executive Officer & Director   
ABM Industries          ...  President, Chief Executive Officer & Director   
ADP                     ...  President, Chief Executive Officer & Director   
AECOM                   ...             Chairman & Chief Executive Officer   

                                                               Address Ticker  \
Title                                                                           
3M                                       3M Center, St. Paul, MN 55144    MMM   
A-Mark Precious Metals  429 Santa Monica Blvd., Santa Monica, CA 90401   AMRK   
ABM Industries                     1 Liberty Plaza, New York, NY 10006    ABM   
ADP                                    1 ADP Blvd., Roseland, NJ 07068    ADP   
AECOM                    1999 Ave. of the Stars, Los Angeles, CA 90067    ACM   

                                               Fullname Revenues Revchange  \
Title                                                                        
3M                                           3M Company    30109      -0.5   
A-Mark Precious Metals     A-Mark Precious Metals, Inc.     6784      11.8   
ABM Industries              ABM Industries Incorporated     5145      -2.8   
ADP                     Automatic Data Processing, Inc.    11668       1.7   
AECOM                                             AECOM    17411      -3.2   

                        Profits  Prftchange  Assets Totshequity  
Title                                                            
3M                       5050.0         4.5   32906     10298.0  
A-Mark Precious Metals      9.3        31.5     437        63.0  
ABM Industries             57.2         -25    2281       974.0  
ADP                      1492.5         2.8   43670      4482.0  
AECOM                      96.

In [124]:
fortune.loc["X":"Y"]

Rank                    Website  Employees          Sector  \
Title                                                                       
XPO Logistics   191         http://www.xpo.com      87000  Transportation   
Xcel Energy     256  http://www.xcelenergy.com      11476          Energy   
Xerox           162       http://www.xerox.com     133600      Technology   

                                   Industry       Hqlocation  \
Title                                                          
XPO Logistics  Transportation and Logistics    Greenwich, CT   
Xcel Energy     Utilities: Gas and Electric  Minneapolis, MN   
Xerox                  Computer Peripherals      Norwalk, CT   

                          Hqaddr       Hqcity Hqstate  Hqzip  ...  \
Title                                                         ...   
XPO Logistics    5 American Lane    Greenwich      CT   6831  ...   
Xcel Energy    414 Nicollet Mall  Minneapolis      MN  55401  ...   
Xerox             45 Glover Ave.      Norwalk      CT   6850  ...   

                                                   Ceo-title  \
Title                                                          
XPO Logistics             Chairman & Chief Executive Officer   
Xcel Energy    Chairman, President & Chief Executive Officer   
Xerox                     Chief Executive Officer & Director   

                                                Address Ticker  \
Title                                                            
XPO Logistics      5 American Lane, Greenwich, CT 06831    XPO   
Xcel Energy    414 Nicollet Mall, Minneapolis, MN 55401    XEL   
Xerox                 45 Glover Ave., Norwalk, CT 06850    XRX   

                          Fullname Revenues Revchange  Profits  Prftchange  \
Title                                                                        
XPO Logistics  XPO Logistics, Inc.    14619      91.8     69.0      -178.2   
Xcel Energy       Xcel Energy Inc.    11107       0.7   1123.4        14.1   
Xerox            Xerox Corporation    17126      -8.2   -477.0      -200.6   

               Assets Totshequity  
Title                              
XPO Logistics   11698      2700.0  
Xcel Energy     41155     11021.0  
Xerox           18145      4803.0  

[3 rows x 22 columns]

## Operacje na kolumnach tekstowych

In [125]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import math as math

In [126]:
frame = pd.read_csv('./data/PublicTransitExpenses.csv', usecols=['Agency', 'Reporter Type', 'Total Operating Expenses'])

In [127]:
frame.head()

Agency     Reporter Type  \
0      Washington County Commissioners  Reduced Reporter   
1      Washington County Commissioners  Reduced Reporter   
2  Texoma Area Paratransit System, Inc     Full Reporter   
3            Kalispel Tribe of Indians  Reduced Reporter   
4            Kalispel Tribe of Indians  Reduced Reporter   

  Total Operating Expenses  
0               $122524.00  
1               $272715.00  
2                 $7295.00  
3                $37416.00  
4               $345789.00

### .str.contains()
sprawdza, czy dany napis występuje w przeszukiwanej kolumnie 

In [129]:
frame["Agency"].str.contains('Washington')                # ważna wielkość liter

0         True
1         True
2        False
3        False
4        False
         ...  
17839    False
17840    False
17841    False
17842    False
17843    False
Name: Agency, Length: 17844, dtype: bool

Żeby otrzymać konkretne wyniki zamiast DS True/False wszystko wkładamy w [] do df

In [130]:
frame[frame["Agency"].str.contains('Washington')]

Agency     Reporter Type  \
0          Washington County Commissioners  Reduced Reporter   
1          Washington County Commissioners  Reduced Reporter   
199               Washington State Ferries     Full Reporter   
213               Washington State Ferries     Full Reporter   
364               Washington State Ferries     Full Reporter   
...                                    ...               ...   
15871  Washington Hancock Community Agency    Rural Reporter   
16213                   City of Washington    Rural Reporter   
16313                   City of Washington    Rural Reporter   
16316   Washington Parish Council on Aging    Rural Reporter   
16780   Washington Parish Council on Aging    Rural Reporter   

      Total Operating Expenses  
0                   $122524.00  
1                   $272715.00  
199               $16215777.00  
213               $11132493.00  
364               $11750411.00  
...                        ...  
15871               $995395.00  
16213               $105117.00  
16313               $102799.00  
16316               $546300.00  
16780               $550224.00  

[137 rows x 3 columns]

### .str.lower()
zmienia wszystkie dane w kolumnie na małe litery

In [132]:
frame["Agency"].str.lower() 

0            washington county commissioners
1            washington county commissioners
2        texoma area paratransit system, inc
3                  kalispel tribe of indians
4                  kalispel tribe of indians
                        ...                 
17839      cedar area transportation service
17840                          tehama county
17841                         city of dixon 
17842        morongo basin transit authority
17843                          city of arvin
Name: Agency, Length: 17844, dtype: object

Żeby zapisać na stałe - przypisujemy do zmiennej

In [133]:
frame["Agency"] = frame["Agency"].str.lower() 

In [134]:
frame.head()

Agency     Reporter Type  \
0      washington county commissioners  Reduced Reporter   
1      washington county commissioners  Reduced Reporter   
2  texoma area paratransit system, inc     Full Reporter   
3            kalispel tribe of indians  Reduced Reporter   
4            kalispel tribe of indians  Reduced Reporter   

  Total Operating Expenses  
0               $122524.00  
1               $272715.00  
2                 $7295.00  
3                $37416.00  
4               $345789.00

In [135]:
frame["Agency"].str.contains('washington')

0         True
1         True
2        False
3        False
4        False
         ...  
17839    False
17840    False
17841    False
17842    False
17843    False
Name: Agency, Length: 17844, dtype: bool

Żeby nie zmieniać danych na stałe możemy dalej dopisywać kolejne funkcje 

In [137]:
frame[frame["Agency"].str.lower().str.endswith('ferries')]

Agency  Reporter Type Total Operating Expenses
199   washington state ferries  Full Reporter             $16215777.00
213   washington state ferries  Full Reporter             $11132493.00
364   washington state ferries  Full Reporter             $11750411.00
397   washington state ferries  Full Reporter             $22400732.00
604   washington state ferries  Full Reporter             $17715787.00
634   washington state ferries  Full Reporter            $236320045.00
653   washington state ferries  Full Reporter            $184425025.00
684   washington state ferries  Full Reporter            $186344635.00
915   washington state ferries  Full Reporter            $236711555.00
1114  washington state ferries  Full Reporter             $23046740.00

Natomiast taki zapis nie jest przejrzysty i zaleca się tworzenie nowych zmiennych

In [138]:
endsWithFerries = frame['Agency'].str.lower().str.strip().str.endswith('ferries')    # strip usuwa spacje z przodu i tyłu

In [139]:
frame[endsWithFerries]

Agency  Reporter Type Total Operating Expenses
199   washington state ferries  Full Reporter             $16215777.00
213   washington state ferries  Full Reporter             $11132493.00
364   washington state ferries  Full Reporter             $11750411.00
397   washington state ferries  Full Reporter             $22400732.00
604   washington state ferries  Full Reporter             $17715787.00
634   washington state ferries  Full Reporter            $236320045.00
653   washington state ferries  Full Reporter            $184425025.00
684   washington state ferries  Full Reporter            $186344635.00
915   washington state ferries  Full Reporter            $236711555.00
1114  washington state ferries  Full Reporter             $23046740.00

#### praca z indexami tekstowymi

In [141]:
frame.set_index('Agency', inplace=True)

In [142]:
frame.head()

Reporter Type Total Operating Expenses
Agency                                                                        
washington county commissioners      Reduced Reporter               $122524.00
washington county commissioners      Reduced Reporter               $272715.00
texoma area paratransit system, inc     Full Reporter                 $7295.00
kalispel tribe of indians            Reduced Reporter                $37416.00
kalispel tribe of indians            Reduced Reporter               $345789.00

In [143]:
frame.index = frame.index.str.strip().str.upper()

In [144]:
frame.head()

Reporter Type Total Operating Expenses
Agency                                                                        
WASHINGTON COUNTY COMMISSIONERS      Reduced Reporter               $122524.00
WASHINGTON COUNTY COMMISSIONERS      Reduced Reporter               $272715.00
TEXOMA AREA PARATRANSIT SYSTEM, INC     Full Reporter                 $7295.00
KALISPEL TRIBE OF INDIANS            Reduced Reporter                $37416.00
KALISPEL TRIBE OF INDIANS            Reduced Reporter               $345789.00


### .split(' ', expand=True)
dzielenie wartości z jednej kolumny na dwie kolumny 

In [145]:
frame["Reporter Type"].str.split(' ').head()      # utworzy listy w miejscu danych

Agency
WASHINGTON COUNTY COMMISSIONERS        [Reduced, Reporter]
WASHINGTON COUNTY COMMISSIONERS        [Reduced, Reporter]
TEXOMA AREA PARATRANSIT SYSTEM, INC       [Full, Reporter]
KALISPEL TRIBE OF INDIANS              [Reduced, Reporter]
KALISPEL TRIBE OF INDIANS              [Reduced, Reporter]
Name: Reporter Type, dtype: object

In [146]:
frame["Reporter Type"].str.split(' ').str[0].head()      # zwraca tylko zerową pozycję z utworzonej listy

Agency
WASHINGTON COUNTY COMMISSIONERS        Reduced
WASHINGTON COUNTY COMMISSIONERS        Reduced
TEXOMA AREA PARATRANSIT SYSTEM, INC       Full
KALISPEL TRIBE OF INDIANS              Reduced
KALISPEL TRIBE OF INDIANS              Reduced
Name: Reporter Type, dtype: object

można tez zapisać za pomocą funkcji .apply()

In [ ]:
frame["Reporter Type"].apply(lambda x: x.split(' ')[0])

In [147]:
frame["Reporter Type"].str.split(' ', expand=True).head()      # Rozdziela na więcej kolumn

0         1
Agency                                                
WASHINGTON COUNTY COMMISSIONERS      Reduced  Reporter
WASHINGTON COUNTY COMMISSIONERS      Reduced  Reporter
TEXOMA AREA PARATRANSIT SYSTEM, INC     Full  Reporter
KALISPEL TRIBE OF INDIANS            Reduced  Reporter
KALISPEL TRIBE OF INDIANS            Reduced  Reporter

In [148]:
frame[["ReporterType1", "ReporterType2"]] = frame["Reporter Type"].str.split(' ', expand=True)     # dodaje kolumny do oryginalnego df

In [149]:
frame.head()

Reporter Type  \
Agency                                                  
WASHINGTON COUNTY COMMISSIONERS      Reduced Reporter   
WASHINGTON COUNTY COMMISSIONERS      Reduced Reporter   
TEXOMA AREA PARATRANSIT SYSTEM, INC     Full Reporter   
KALISPEL TRIBE OF INDIANS            Reduced Reporter   
KALISPEL TRIBE OF INDIANS            Reduced Reporter   

                                    Total Operating Expenses ReporterType1  \
Agency                                                                       
WASHINGTON COUNTY COMMISSIONERS                   $122524.00       Reduced   
WASHINGTON COUNTY COMMISSIONERS                   $272715.00       Reduced   
TEXOMA AREA PARATRANSIT SYSTEM, INC                 $7295.00          Full   
KALISPEL TRIBE OF INDIANS                          $37416.00       Reduced   
KALISPEL TRIBE OF INDIANS                         $345789.00       Reduced   

                                    ReporterType2  
Agency                                             
WASHINGTON COUNTY COMMISSIONERS          Reporter  
WASHINGTON COUNTY COMMISSIONERS          Reporter  
TEXOMA AREA PARATRANSIT SYSTEM, INC      Reporter  
KALISPEL TRIBE OF INDIANS                Reporter  
KALISPEL TRIBE OF INDIANS                Reporter

.split(' ', expand=True, n=5) - n=5 określa ile maksymalnie kolumn stworzyć licząc od 0, czyli n=5 stworzy 6 kolumn.

### wykorzystanie utworzonej funkcji 
### .apply()

In [153]:
def getComment(row):                        # wykonuję funkcję dla każdego wiersza. 
    reporterType = row['Reporter Type']
    cost = float(row['Total Operating Expenses'].replace("$", ""))
    
    if(cost > 200000):
        comment = "CLASS A"
    else:
        comment = "CLASS B"
    return (reporterType + "/" + comment)

In [154]:
frame.apply(getComment, axis=1).head()

Agency
WASHINGTON COUNTY COMMISSIONERS        Reduced Reporter/CLASS B
WASHINGTON COUNTY COMMISSIONERS        Reduced Reporter/CLASS A
TEXOMA AREA PARATRANSIT SYSTEM, INC       Full Reporter/CLASS B
KALISPEL TRIBE OF INDIANS              Reduced Reporter/CLASS B
KALISPEL TRIBE OF INDIANS              Reduced Reporter/CLASS A
dtype: object

## Ćwiczenie 
operacje na kolumnach tekstowych

In [2]:
import pandas as pd 
import numpy as np

In [3]:
fortune = pd.read_csv("./data/Fortune_500_2017.csv", usecols=['Rank','Title','Industry','Hqlocation','Sector'])

In [4]:
fortune.head()

Rank               Title       Sector  \
0     1             Walmart    Retailing   
1     2  Berkshire Hathaway   Financials   
2     3               Apple   Technology   
3     4         Exxon Mobil       Energy   
4     5            McKesson  Wholesalers   

                                   Industry         Hqlocation  
0                     General Merchandisers    Bentonville, AR  
1  Insurance: Property and Casualty (Stock)          Omaha, NE  
2               Computers, Office Equipment      Cupertino, CA  
3                        Petroleum Refining         Irving, TX  
4                  Wholesalers: Health Care  San Francisco, CA

In [5]:
fortune['Sector'].str.upper().head()

0      RETAILING
1     FINANCIALS
2     TECHNOLOGY
3         ENERGY
4    WHOLESALERS
Name: Sector, dtype: object

In [6]:
fortune[fortune['Industry'].str.lower().str.contains("comp")]

Rank                Title      Sector  \
2       3                Apple  Technology   
27     28            Microsoft  Technology   
40     41    Dell Technologies  Technology   
46     47                Intel  Technology   
60     61                   HP  Technology   
80     81               Oracle  Technology   
118   119             Qualcomm  Technology   
151   152        Jabil Circuit  Technology   
161   162                Xerox  Technology   
205   206    Texas Instruments  Technology   
216   217      Western Digital  Technology   
225   226    Micron Technology  Technology   
264   265    Applied Materials  Technology   
325   326       salesforce.com  Technology   
386   387               Nvidia  Technology   
405   406  Activision Blizzard  Technology   
408   409                  NCR  Technology   
413   414              Sanmina  Technology   
439   440         Lam Research  Technology   
442   443        Adobe Systems  Technology   
464   465             Symantec  Technology   
467   468               NetApp  Technology   

                                           Industry          Hqlocation  
2                       Computers, Office Equipment       Cupertino, CA  
27                                Computer Software         Redmond, WA  
40                      Computers, Office Equipment      Round Rock, TX  
46   Semiconductors and Other Electronic Components     Santa Clara, CA  
60                      Computers, Office Equipment       Palo Alto, CA  
80                                Computer Software    Redwood City, CA  
118  Semiconductors and Other Electronic Components       San Diego, CA  
151  Semiconductors and Other Electronic Components  St. Petersburg, FL  
161                            Computer Peripherals         Norwalk, CT  
205  Semiconductors and Other Electronic Components          Dallas, TX  
216                            Computer Peripherals          Irvine, CA  
225  Semiconductors and Other Electronic Components           Boise, ID  
264  Semiconductors and Other Electronic Components     Santa Clara, CA  
325                               Computer Software   San Francisco, CA  
386  Semiconductors and Other Electronic Components     Santa Clara, CA  
405                               Computer Software    Santa Monica, CA  
408                     Computers, Office Equipment          Duluth, GA  
413  Semiconductors and Other Electronic Components        San Jose, CA  
439  Semiconductors and Other Electronic Components         Fremont, CA  
442                               Computer Software        San Jose, CA  
464                               Computer Software   Mountain View, CA  
467                            Computer Peripherals       Sunnyvale, CA

In [162]:
fortune[["City", "State"]] = fortune['Hqlocation'].str.strip().str.split(",", expand=True)

In [163]:
fortune.head()

Rank               Title       Sector  \
0     1             Walmart    Retailing   
1     2  Berkshire Hathaway   Financials   
2     3               Apple   Technology   
3     4         Exxon Mobil       Energy   
4     5            McKesson  Wholesalers   

                                   Industry         Hqlocation           City  \
0                     General Merchandisers    Bentonville, AR    Bentonville   
1  Insurance: Property and Casualty (Stock)          Omaha, NE          Omaha   
2               Computers, Office Equipment      Cupertino, CA      Cupertino   
3                        Petroleum Refining         Irving, TX         Irving   
4                  Wholesalers: Health Care  San Francisco, CA  San Francisco   

  State  
0    AR  
1    NE  
2    CA  
3    TX  
4    CA

In [7]:
def BuildShortcut(row):
    industry = row["Industry"]
    result =""
    for i in industry.split(" "):
        result += i[0]
    
    return result

In [8]:
fortune.apply(BuildShortcut, axis=1)

0         GM
1      IPaC(
2        COE
3         PR
4        WHC
       ...  
495      SRO
496        H
497     ISaR
498        E
499      DOS
Length: 500, dtype: object

In [181]:
BuildShortcut({'Industry':'Factory Under Newspaper'})

'FUN'

In [182]:
fortune["IndustryShort"] = fortune.apply(BuildShortcut, axis=1)

In [183]:
fortune.head()

Rank               Title       Sector  \
0     1             Walmart    Retailing   
1     2  Berkshire Hathaway   Financials   
2     3               Apple   Technology   
3     4         Exxon Mobil       Energy   
4     5            McKesson  Wholesalers   

                                   Industry         Hqlocation           City  \
0                     General Merchandisers    Bentonville, AR    Bentonville   
1  Insurance: Property and Casualty (Stock)          Omaha, NE          Omaha   
2               Computers, Office Equipment      Cupertino, CA      Cupertino   
3                        Petroleum Refining         Irving, TX         Irving   
4                  Wholesalers: Health Care  San Francisco, CA  San Francisco   

  State IndustryShort  
0    AR            GM  
1    NE         IPaC(  
2    CA           COE  
3    TX            PR  
4    CA           WHC

## Przygotowanie danych po imporcie

In [184]:
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt

In [185]:
frame = pd.read_csv("./data/PublicTransitExpenses.csv", low_memory=False)

In [186]:
frame.head()

5 digit NTD ID 4 digit NTD ID                               Agency  \
0          30098           3098      Washington County Commissioners   
1          30098           3098      Washington County Commissioners   
2          60107           6107  Texoma Area Paratransit System, Inc   
3              9           0T09            Kalispel Tribe of Indians   
4              9           0T09            Kalispel Tribe of Indians   

      Reporter Type Subrecipient Type Organization Type  \
0  Reduced Reporter               NaN               NaN   
1  Reduced Reporter               NaN               NaN   
2     Full Reporter               NaN               NaN   
3  Reduced Reporter               NaN             Tribe   
4  Reduced Reporter               NaN             Tribe   

   2015 Total Mode Vehicles (VOMS) Mode Type of Service  Primary UZA Code  \
0                              NaN   DR              PT               NaN   
1                              NaN   MB              PT               NaN   
2                              NaN   CB              PT               NaN   
3                              4.0   DR              DO               0.0   
4                              4.0   MB              DO               0.0   

   ... Casualty and Liability Costs  Taxes  PT Funds In Report  \
0  ...                        $0.00  $0.00               $0.00   
1  ...                        $0.00  $0.00               $0.00   
2  ...                          NaN    NaN            $3398.00   
3  ...                          NaN    NaN                 NaN   
4  ...                          NaN    NaN                 NaN   

   PT Funds Reported Separately  Miscellaneous Expenses  \
0                         $0.00                   $0.00   
1                         $0.00                   $0.00   
2                           NaN                     NaN   
3                           NaN                     NaN   
4                           NaN                     NaN   

   Reduced Reporter Total OE  Total Operating Expenses  \
0                 $122524.00                $122524.00   
1                 $272715.00                $272715.00   
2                        NaN                  $7295.00   
3                  $37416.00                 $37416.00   
4                 $345789.00                $345789.00   

   Total Operating Expenses (No Funds Reported Separately)  \
0                                         $122524.00         
1                                         $272715.00         
2                                           $7295.00         
3                                          $37416.00         
4                                         $345789.00         

  ADA Related Expenses Location 1  
0                $0.00        NaN  
1                $0.00        NaN  
2                  NaN        NaN  
3                  NaN  Usk, WA\n  
4                  NaN  Usk, WA\n  

[5 rows x 39 columns]

In [187]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 39 columns):
 #   Column                                                   Non-Null Count  Dtype  
---  ------                                                   --------------  -----  
 0   5 digit NTD ID                                           17844 non-null  object 
 1   4 digit NTD ID                                           17719 non-null  object 
 2   Agency                                                   17844 non-null  object 
 3   Reporter Type                                            17844 non-null  object 
 4   Subrecipient Type                                        3072 non-null   object 
 5   Organization Type                                        17759 non-null  object 
 6   2015 Total Mode Vehicles (VOMS)                          17522 non-null  float64
 7   Mode                                                     17844 non-null  object 
 8   Type of Service           

Zdarza się tak, że Python nie jest w stanie określic jaki jest typ danej kolumny i przypisuje jej typ object, przez co wykorzystywane jest więcej pamięci. W takich przypadkach do załadowania df potrzebujemy parametru low_memory=False.
<br> Żeby zmniejszyć ilość wykorzystywanej pamięci zaciągajmy tylko te kolumny, których potrzebujemy do analizy

In [189]:
frame = pd.read_csv('./data/PublicTransitExpenses.csv', usecols=["Agency", "Reporter Type", "Organization Type", 
                                                                 "Rail (Y/N)", "Fixed Route (Y/N)", "Service Costs", 
                                                                'Tires and Tubes', "Total Operating Expenses", 
                                                                 "Service Area Population"], low_memory=False)

In [190]:
frame.head()

Agency     Reporter Type Organization Type  \
0      Washington County Commissioners  Reduced Reporter               NaN   
1      Washington County Commissioners  Reduced Reporter               NaN   
2  Texoma Area Paratransit System, Inc     Full Reporter               NaN   
3            Kalispel Tribe of Indians  Reduced Reporter             Tribe   
4            Kalispel Tribe of Indians  Reduced Reporter             Tribe   

   Service Area Population Rail (Y/N) Fixed Route (Y/N) Service Costs  \
0                      NaN        NaN               NaN         $0.00   
1                      NaN        NaN               NaN         $0.00   
2                      NaN        NaN               NaN           NaN   
3                      NaN          N                 N           NaN   
4                      NaN          N                 Y           NaN   

  Tires and Tubes Total Operating Expenses  
0           $0.00               $122524.00  
1           $0.00               $272715.00  
2             NaN                 $7295.00  
3             NaN                $37416.00  
4             NaN               $345789.00

In [193]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Agency                    17844 non-null  object 
 1   Reporter Type             17844 non-null  object 
 2   Organization Type         17759 non-null  object 
 3   Service Area Population   14418 non-null  float64
 4   Rail (Y/N)                17318 non-null  object 
 5   Fixed Route (Y/N)         17318 non-null  object 
 6   Service Costs             10262 non-null  object 
 7   Tires and Tubes           5615 non-null   object 
 8   Total Operating Expenses  17844 non-null  object 
dtypes: float64(1), object(8)
memory usage: 9.5 MB


Samo zaczytanie tylko wybranych kolumn zmniejszyło wykorzystywaną pamięc z 30MB do 9MB
<br> Jeśli chcemy to możemy zmienić nazwy kolumn, żeby się prościej pracowało

In [195]:
newColumnNames = {"Agency"                  :"Agency",
                  "Reporter Type"           :"ReporterType",
                  "Organization Type"       :"OrgType",
                  "Rail (Y/N)"              :"IsRail",
                  "Fixed Route (Y/N)"       :"IsFixedRoute",
                  "Service Costs"           :"ServiceCost",
                  'Tires and Tubes'         :"TiresTubesCost",
                  "Total Operating Expenses":"TotalExpenses",
                  "Service Area Population" :"Population"}

In [196]:
frame.rename(columns=newColumnNames, inplace=True)

In [197]:
frame.head()

Agency      ReporterType OrgType  Population  \
0      Washington County Commissioners  Reduced Reporter     NaN         NaN   
1      Washington County Commissioners  Reduced Reporter     NaN         NaN   
2  Texoma Area Paratransit System, Inc     Full Reporter     NaN         NaN   
3            Kalispel Tribe of Indians  Reduced Reporter   Tribe         NaN   
4            Kalispel Tribe of Indians  Reduced Reporter   Tribe         NaN   

  IsRail IsFixedRoute ServiceCost TiresTubesCost TotalExpenses  
0    NaN          NaN       $0.00          $0.00    $122524.00  
1    NaN          NaN       $0.00          $0.00    $272715.00  
2    NaN          NaN         NaN            NaN      $7295.00  
3      N            N         NaN            NaN     $37416.00  
4      N            Y         NaN            NaN    $345789.00

Sprawdzamy, czy warto konwertować kolumnę do typu category

In [198]:
frame['ReporterType'].nunique()

4

In [199]:
frame['ReporterType'].count()

17844

Kolumna ReporterType posiada 17844 elementów, które są uzupełnione 4 różnymi wartościami - Idealna do konwersji na typ category

In [200]:
frame['ReporterType'].value_counts()

Full Reporter       13345
Rural Reporter       3116
Reduced Reporter     1313
Separate Service       70
Name: ReporterType, dtype: int64

In [201]:
frame['ReporterType'] = frame['ReporterType'].astype('category')

In [202]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Agency          17844 non-null  object  
 1   ReporterType    17844 non-null  category
 2   OrgType         17759 non-null  object  
 3   Population      14418 non-null  float64 
 4   IsRail          17318 non-null  object  
 5   IsFixedRoute    17318 non-null  object  
 6   ServiceCost     10262 non-null  object  
 7   TiresTubesCost  5615 non-null   object  
 8   TotalExpenses   17844 non-null  object  
dtypes: category(1), float64(1), object(7)
memory usage: 8.3 MB


Zmniejszyliśmy wykorzystaną pamięć o kolejny MB

In [203]:
frame['Agency'] = frame['Agency'].astype('category')

In [204]:
frame['OrgType'] = frame['OrgType'].astype('category')

In [205]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Agency          17844 non-null  category
 1   ReporterType    17844 non-null  category
 2   OrgType         17759 non-null  category
 3   Population      14418 non-null  float64 
 4   IsRail          17318 non-null  object  
 5   IsFixedRoute    17318 non-null  object  
 6   ServiceCost     10262 non-null  object  
 7   TiresTubesCost  5615 non-null   object  
 8   TotalExpenses   17844 non-null  object  
dtypes: category(3), float64(1), object(5)
memory usage: 5.1 MB


Już spadło do tylko 5MB
<br> Następnie sprawdzamy kolumnę Population, która powinna być Int, a jest Float(dlatego, że przyjmuje wartości NaN)
<br> Więc zmieniamy NaN na 0 i konwertujemy kolumne na Int 

In [206]:
frame["Population"].fillna(0, inplace=True)

In [207]:
frame['Population'] = frame['Population'].astype('int')

In [208]:
frame[frame['Population']>0].head()

Agency      ReporterType  \
71            Reno-Sparks Indian Colony  Reduced Reporter   
92                    City of Pocatello  Reduced Reporter   
93   Lee-Russell Council of Governments  Reduced Reporter   
94    Aiken Area Council on Aging, Inc.  Reduced Reporter   
101                    City of Lawrence     Full Reporter   

                                               OrgType  Population IsRail  \
71                                               Tribe        1127      N   
92   City, County or Local Government Unit or Depar...       81730      N   
93                   MPO, COG or Other Planning Agency      193194      N   
94                                Area Agency on Aging      160099      N   
101                              Consolidated Reporter       87643      N   

    IsFixedRoute ServiceCost TiresTubesCost TotalExpenses  
71             Y         NaN            NaN    $174810.00  
92             Y         NaN            NaN   $1119899.00  
93             N         NaN            NaN   $1540633.00  
94             Y         NaN            NaN    $525325.00  
101            Y         NaN            NaN    $937407.00

In [209]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Agency          17844 non-null  category
 1   ReporterType    17844 non-null  category
 2   OrgType         17759 non-null  category
 3   Population      17844 non-null  int32   
 4   IsRail          17318 non-null  object  
 5   IsFixedRoute    17318 non-null  object  
 6   ServiceCost     10262 non-null  object  
 7   TiresTubesCost  5615 non-null   object  
 8   TotalExpenses   17844 non-null  object  
dtypes: category(3), int32(1), object(5)
memory usage: 5.1 MB


Teraz możemy zmienić kolumny, które posiadają "Y" i "N" tak, żeby były wartościami True i False

In [210]:
frame["IsRail"].replace(("Y", "N"),(True,False), inplace=True)

In [211]:
frame["IsFixedRoute"].replace(("Y", "N"),(True,False), inplace=True)

In [212]:
frame.head()

Agency      ReporterType OrgType  Population  \
0      Washington County Commissioners  Reduced Reporter     NaN           0   
1      Washington County Commissioners  Reduced Reporter     NaN           0   
2  Texoma Area Paratransit System, Inc     Full Reporter     NaN           0   
3            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   
4            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   

  IsRail IsFixedRoute ServiceCost TiresTubesCost TotalExpenses  
0    NaN          NaN       $0.00          $0.00    $122524.00  
1    NaN          NaN       $0.00          $0.00    $272715.00  
2    NaN          NaN         NaN            NaN      $7295.00  
3  False        False         NaN            NaN     $37416.00  
4  False         True         NaN            NaN    $345789.00

Żeby zmienić typ kolumny na bool nie możemy mieć wartości NaN, dlatego musimy je zmienić na coś innego

In [213]:
frame["IsRail"].fillna(False, inplace=True)          # zmieniamy wszystkie wartości NaN na False - niezawsze jest to dobra decyzja

In [214]:
frame["IsFixedRoute"].fillna(False, inplace=True) 

In [215]:
frame["IsRail"] = frame["IsRail"].astype('bool')

In [216]:
frame["IsFixedRoute"] = frame["IsFixedRoute"].astype('bool')

In [217]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Agency          17844 non-null  category
 1   ReporterType    17844 non-null  category
 2   OrgType         17759 non-null  category
 3   Population      17844 non-null  int32   
 4   IsRail          17844 non-null  bool    
 5   IsFixedRoute    17844 non-null  bool    
 6   ServiceCost     10262 non-null  object  
 7   TiresTubesCost  5615 non-null   object  
 8   TotalExpenses   17844 non-null  object  
dtypes: bool(2), category(3), int32(1), object(3)
memory usage: 3.2 MB


Teraz wykorzystanie pamięci wynosi 3.2MB

In [218]:
frame.head()

Agency      ReporterType OrgType  Population  \
0      Washington County Commissioners  Reduced Reporter     NaN           0   
1      Washington County Commissioners  Reduced Reporter     NaN           0   
2  Texoma Area Paratransit System, Inc     Full Reporter     NaN           0   
3            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   
4            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   

   IsRail  IsFixedRoute ServiceCost TiresTubesCost TotalExpenses  
0   False         False       $0.00          $0.00    $122524.00  
1   False         False       $0.00          $0.00    $272715.00  
2   False         False         NaN            NaN      $7295.00  
3   False         False         NaN            NaN     $37416.00  
4   False          True         NaN            NaN    $345789.00

Zmieniamy kolumny, ktore mają  wrtości ze znakiem $

In [220]:
frame['ServiceCost'] = frame['ServiceCost'].strlace("$", "")

In [235]:
frame['TiresTubesCost'] = frame['TiresTubesCost'].str.replace("$", " ")

C:\Users\hp\AppData\Local\Temp\ipykernel_4448\3468894161.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  frame['TiresTubesCost'] = frame['TiresTubesCost'].str.replace("$", " ")


In [229]:
frame['TotalExpenses'] = frame['TotalExpenses'].str.replace("$", "")

In [236]:
frame.head()

Agency      ReporterType OrgType  Population  \
0      Washington County Commissioners  Reduced Reporter     NaN           0   
1      Washington County Commissioners  Reduced Reporter     NaN           0   
2  Texoma Area Paratransit System, Inc     Full Reporter     NaN           0   
3            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   
4            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   

   IsRail  IsFixedRoute ServiceCost TiresTubesCost TotalExpenses  
0   False         False        0.00           0.00     122524.00  
1   False         False        0.00           0.00     272715.00  
2   False         False         NaN            NaN       7295.00  
3   False         False         NaN            NaN      37416.00  
4   False          True         NaN            NaN     345789.00

In [237]:
frame['ServiceCost'] = frame['ServiceCost'].astype('float')

In [240]:
frame['TiresTubesCost'] = frame['TiresTubesCost'].astype('float')

In [239]:
frame['TotalExpenses'] = frame['TotalExpenses'].astype('float')

In [241]:
frame.head()

Agency      ReporterType OrgType  Population  \
0      Washington County Commissioners  Reduced Reporter     NaN           0   
1      Washington County Commissioners  Reduced Reporter     NaN           0   
2  Texoma Area Paratransit System, Inc     Full Reporter     NaN           0   
3            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   
4            Kalispel Tribe of Indians  Reduced Reporter   Tribe           0   

   IsRail  IsFixedRoute  ServiceCost  TiresTubesCost  TotalExpenses  
0   False         False          0.0             0.0       122524.0  
1   False         False          0.0             0.0       272715.0  
2   False         False          NaN             NaN         7295.0  
3   False         False          NaN             NaN        37416.0  
4   False          True          NaN             NaN       345789.0

In [242]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Agency          17844 non-null  category
 1   ReporterType    17844 non-null  category
 2   OrgType         17759 non-null  category
 3   Population      17844 non-null  int32   
 4   IsRail          17844 non-null  bool    
 5   IsFixedRoute    17844 non-null  bool    
 6   ServiceCost     10262 non-null  float64 
 7   TiresTubesCost  5615 non-null   float64 
 8   TotalExpenses   17844 non-null  float64 
dtypes: bool(2), category(3), float64(3), int32(1)
memory usage: 846.4 KB


Zmniejszyliśmy zużycie pamięci do 850KB z początkowych 30MB
<br> Opcjonalnie możemy zmienić też format pisowni w niektórych kolumnach na bardziej przejrzysty i zgodny z przyjętymi standardami

In [243]:
frame['Agency'] = frame['Agency'].str.title().astype('category')

In [244]:
frame['ReporterType'] = frame['ReporterType'].str.upper().astype('category')

In [245]:
frame['OrgType'] = frame['OrgType'].str.upper().astype('category')

In [246]:
frame.head()

Agency      ReporterType OrgType  Population  \
0      Washington County Commissioners  REDUCED REPORTER     NaN           0   
1      Washington County Commissioners  REDUCED REPORTER     NaN           0   
2  Texoma Area Paratransit System, Inc     FULL REPORTER     NaN           0   
3            Kalispel Tribe Of Indians  REDUCED REPORTER   TRIBE           0   
4            Kalispel Tribe Of Indians  REDUCED REPORTER   TRIBE           0   

   IsRail  IsFixedRoute  ServiceCost  TiresTubesCost  TotalExpenses  
0   False         False          0.0             0.0       122524.0  
1   False         False          0.0             0.0       272715.0  
2   False         False          NaN             NaN         7295.0  
3   False         False          NaN             NaN        37416.0  
4   False          True          NaN             NaN       345789.0

In [247]:
frame.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17844 entries, 0 to 17843
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Agency          17844 non-null  category
 1   ReporterType    17844 non-null  category
 2   OrgType         17759 non-null  category
 3   Population      17844 non-null  int32   
 4   IsRail          17844 non-null  bool    
 5   IsFixedRoute    17844 non-null  bool    
 6   ServiceCost     10262 non-null  float64 
 7   TiresTubesCost  5615 non-null   float64 
 8   TotalExpenses   17844 non-null  float64 
dtypes: bool(2), category(3), float64(3), int32(1)
memory usage: 846.1 KB


## Ćwizcenie 
Przygotowanie danych

In [1]:
import pandas as pd 
import numpy as np

In [3]:
fb = pd.read_csv('./data/mrbean_facebook_statuses_with_nulls.csv')

In [4]:
fb.head()

status_id  \
0  17774451468_10154154735571469   
1  17774451468_10154146584106469   
2  17774451468_10154135502911469   
3  17774451468_10154138120151469   
4  17774451468_10154135502476469   

                                      status_message  \
0  It's time for Mr Bean and Teddy to get ready f...   
1                                                NaN   
2                                                NaN   
3  Mr Bean is ready to do some shopping but bewar...   
4                                                NaN   

                           link_name status_type  \
0  Mr Bean - Preparing To Go Camping       video   
1                    Timeline Photos       photo   
2                    Timeline Photos       photo   
3                    www.youtube.com       video   
4                    Timeline Photos       photo   

                                         status_link     status_published  \
0                       https://youtu.be/fgURU75gTMQ  2016-06-23 11:35:48   
1  https://www.facebook.com/MrBean/photos/a.10150...  2016-06-21 10:00:01   
2  https://www.facebook.com/MrBean/photos/a.10150...  2016-06-19 10:00:01   
3        https://www.youtube.com/watch?v=58Z8J0PbLS8  2016-06-18 11:00:02   
4  https://www.facebook.com/MrBean/photos/a.10150...  2016-06-17 09:00:01   

   num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0        16570.0         223.0       338.0      16079        166         6   
1       119886.0        2171.0      1657.0     114008        946       104   
2            NaN        2395.0     10050.0     218579       2988       188   
3        20913.0         214.0       257.0      20338        211        20   
4       223602.0         880.0      2228.0     215591       2872       259   

   num_hahas  num_sads  num_angrys  
0        320         0           0  
1       4795        16          17  
2      13992        52          53  
3        336         0           8  
4       4814        27          39

In [5]:
fb.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   status_id         56 non-null     object 
 1   status_message    40 non-null     object 
 2   link_name         56 non-null     object 
 3   status_type       56 non-null     object 
 4   status_link       56 non-null     object 
 5   status_published  56 non-null     object 
 6   num_reactions     54 non-null     float64
 7   num_comments      55 non-null     float64
 8   num_shares        56 non-null     float64
 9   num_likes         56 non-null     int64  
 10  num_loves         56 non-null     int64  
 11  num_wows          56 non-null     int64  
 12  num_hahas         56 non-null     int64  
 13  num_sads          56 non-null     int64  
 14  num_angrys        56 non-null     int64  
dtypes: float64(3), int64(6), object(6)
memory usage: 34.4 KB


In [6]:
fb = pd.read_csv('./data/mrbean_facebook_statuses_with_nulls.csv', usecols=["status_message", "status_type", "link_name",
"num_reactions", "num_shares", "num_likes"])

In [7]:
fb.head()

status_message  \
0  It's time for Mr Bean and Teddy to get ready f...   
1                                                NaN   
2                                                NaN   
3  Mr Bean is ready to do some shopping but bewar...   
4                                                NaN   

                           link_name status_type  num_reactions  num_shares  \
0  Mr Bean - Preparing To Go Camping       video        16570.0       338.0   
1                    Timeline Photos       photo       119886.0      1657.0   
2                    Timeline Photos       photo            NaN     10050.0   
3                    www.youtube.com       video        20913.0       257.0   
4                    Timeline Photos       photo       223602.0      2228.0   

   num_likes  
0      16079  
1     114008  
2     218579  
3      20338  
4     215591

In [8]:
fb.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   status_message  40 non-null     object 
 1   link_name       56 non-null     object 
 2   status_type     56 non-null     object 
 3   num_reactions   54 non-null     float64
 4   num_shares      56 non-null     float64
 5   num_likes       56 non-null     int64  
dtypes: float64(2), int64(1), object(3)
memory usage: 16.1 KB


In [9]:
fb.count()

status_message    40
link_name         56
status_type       56
num_reactions     54
num_shares        56
num_likes         56
dtype: int64

In [10]:
len(fb)

56

In [12]:
fb['status_type'].nunique()

3

In [15]:
fb['status_type'].value_counts()

video    32
photo    22
link      2
Name: status_type, dtype: int64

In [17]:
fb['status_type'] = fb['status_type'].astype('category')

In [19]:
fb['link_name'].nunique()

32

In [18]:
fb['link_name'].value_counts()

Timeline Photos                                                               22
www.youtube.com                                                                4
Mr. Bean - Sneezing and Snoozing In Church                                     1
Mr. Bean - Mystery creature                                                    1
Mr. Bean - Saxophone Player                                                    1
Mr. Bean - Elderly people on stairs                                            1
Rowan Atkinson: I will never wave goodbye to Mr Bean                           1
Mr. Bean's Happy Moments                                                       1
Mr. Bean - At the Disco                                                        1
Mr Bean - Preparing To Go Camping                                              1
Mr. Bean - Shopping for Towels                                                 1
Mr. Bean - Counting Sheep.                                                     1
Mr. Bean - High Dive        

In [20]:
fb['link_name'] = fb['link_name'].astype('category')

In [21]:
fb.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   status_message  40 non-null     object  
 1   link_name       56 non-null     category
 2   status_type     56 non-null     category
 3   num_reactions   54 non-null     float64 
 4   num_shares      56 non-null     float64 
 5   num_likes       56 non-null     int64   
dtypes: category(2), float64(2), int64(1), object(1)
memory usage: 12.4 KB


In [22]:
fb['num_reactions'] = fb['num_reactions'].fillna(0)

In [25]:
fb['num_shares'] = fb['num_shares'].fillna(0)

In [27]:
fb.head(10)

status_message  \
0  It's time for Mr Bean and Teddy to get ready f...   
1                                                NaN   
2                                                NaN   
3  Mr Bean is ready to do some shopping but bewar...   
4                                                NaN   
5            Mr. Bean's sweet treat causes him pain!   
6  Mr Bean's chasing a ray of sunshine!\r\nhttps:...   
7  Mr Bean is enjoying the sunshine....that is un...   
8                                                NaN   
9                                                NaN   

                           link_name status_type  num_reactions  num_shares  \
0  Mr Bean - Preparing To Go Camping       video        16570.0       338.0   
1                    Timeline Photos       photo       119886.0      1657.0   
2                    Timeline Photos       photo            0.0     10050.0   
3                    www.youtube.com       video        20913.0       257.0   
4                    Timeline Photos       photo       223602.0      2228.0   
5             Mr Bean - Broken Tooth       video        24742.0       448.0   
6          Mr Bean - Ray Of Sunshine       video        21886.0       471.0   
7                Mr Bean - Sun Block       video        22015.0       546.0   
8                    Timeline Photos       photo       196913.0      9849.0   
9                    Timeline Photos       photo       196366.0      5378.0   

   num_likes  
0      16079  
1     114008  
2     218579  
3      20338  
4     215591  
5      23977  
6      21246  
7      21293  
8     190476  
9     190857

In [28]:
fb.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   status_message  40 non-null     object  
 1   link_name       56 non-null     category
 2   status_type     56 non-null     category
 3   num_reactions   56 non-null     float64 
 4   num_shares      56 non-null     float64 
 5   num_likes       56 non-null     int64   
dtypes: category(2), float64(2), int64(1), object(1)
memory usage: 12.4 KB


In [29]:
fb['num_reactions'] = fb['num_reactions'].astype('int')

In [31]:
fb['num_shares'] = fb['num_shares'].astype('int')

In [30]:
fb['num_likes'] = fb['num_likes'].astype('int')

In [32]:
fb.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   status_message  40 non-null     object  
 1   link_name       56 non-null     category
 2   status_type     56 non-null     category
 3   num_reactions   56 non-null     int32   
 4   num_shares      56 non-null     int32   
 5   num_likes       56 non-null     int32   
dtypes: category(2), int32(3), object(1)
memory usage: 11.8 KB
